src/main/java/org/optaplanner/springboottaskassigning/solver/SolverManager.java

```diff
@@ -16,16 +16,21 @@
 
 package org.optaplanner.springboottaskassigning.solver;
 
+import java.util.function.Consumer;
+
 import org.optaplanner.core.api.score.Score;
 
 public interface SolverManager<Solution_> {
 
-    // TODO iterate over planning solution entities and set tenantId, or induce it from solution.getTenantId()
-    void solve(Comparable<?> tenantId, Solution_ planningSolution);
+    // FIXME CRITICAL BUG: iterate over planning solution entities and set tenantId, or induce it from solution.getTenantId()
+    void solve(Comparable<?> tenantId, Solution_ planningSolution,
```

Comment:
The API requiring `tenantId` for each operation is not natural for embedded use case - one would rather expect getting a reference to the solving that has been started, or passing a consumer that the `SolverManager` calls once a new best solution is ready.

Another question is, whether to run a thread executor in this per-client reference or in the manager itself (and share it for different datasets).
Rather than sharing a thread executor among different tenants, we should aim to create a new pod for each tenant. Thus I would keep the Java API simple and designed for a single tenant.

These are questions to think about when you move the `SolverManager` from this example to oplanner-core - no need to implement it now in this example.

Reply:
For the first question, the idea is to not leak the internal objects by passing a reference (to the `SolverTask` for example), so the other option is returning an ID from `SolverManager.solve()` to be used for other operations, which is same as using `tenantId`. Adding a consumer limits the notifications only to when a new bestSolution is found (so if user wants to get the solution after solving ended they'd need to save it somewhere first), which is already implemented but in this case users can only get the data from a data_source or their application instead of from SolverManager, which IMO is better than having duplicate data (both in SolverManager and in data_source), i.e. `SolverManager` discards of all data of a specific tenant when solving ends, but this forces them to implement `onSolvingEnded` listener and persist the solution. @ge0ffrey wdyt?

The second question is more interesting, this is going to be another implementation (i.e. `DistributedSolverManager`) and tenants need not worry about which one is used in the backend, but `SolverManager` still needs to know which tenant is asking for `solverStatus` for example. Please let us know of any suggestion or simplifications you can think of, since we're still changing the API before moving it to optaplanner-core.